src/transformers/modeling_reformer.py

```diff
@@ -0,0 +1,1757 @@
+            device = input_ids.device
+        elif inputs_embeds is not None:
+            input_shape = inputs_embeds.size()[:-1]  # noqa: F841
+            device = inputs_embeds.device
+        else:
+            raise ValueError("You have to specify either input_ids or inputs_embeds")
+
+        assert (
+            len(input_shape) == 2
+        ), "`input_ids` have be of shape `[batch_size, sequence_length]`, but got shape: {}".format(input_shape)
+
+        # prepare head mask
+        head_mask = self.get_head_mask(head_mask, self.config.num_hidden_layers, is_attention_chunked=True)
+
+        # original sequence length for padding
+        orig_sequence_length = input_shape[-1]
+
+        # if needs padding
+        least_common_mult_chunk_length = _get_least_common_mult_chunk_len(self.config)
+        has_to_pad_to_match_chunk_length = input_shape[-1] % least_common_mult_chunk_length != 0
+
+        if has_to_pad_to_match_chunk_length:
+            # pad input
+            input_ids, inputs_embeds, attention_mask, position_ids, input_shape = self._pad_to_mult_of_chunk_length(
+                input_ids,
+                inputs_embeds,
+                attention_mask,
+                position_ids,
+                input_shape,
+                least_common_mult_chunk_length,
+                device,
+            )
+
+        embedding_output = self.embeddings(input_ids=input_ids, position_ids=position_ids, inputs_embeds=inputs_embeds)
+
+        encoder_outputs = self.encoder(
+            hidden_states=embedding_output,
+            head_mask=head_mask,
+            attention_mask=attention_mask,
+            num_hashes=num_hashes,
+            do_output_hidden_states=do_output_hidden_states,
+            do_output_attentions=do_output_attentions,
+        )
+        sequence_output = encoder_outputs.hidden_states
+
+        # if padding was applied
+        if has_to_pad_to_match_chunk_length:
+            sequence_output = sequence_output[:, :orig_sequence_length]
+
+        outputs = (sequence_output,)
+        # TODO(PVP): Replace by named tuple after namedtuples are introduced in the library.
+        if do_output_hidden_states is True:
+            outputs = outputs + (encoder_outputs.all_hidden_states,)
+        if do_output_attentions is True:
+            outputs = outputs + (encoder_outputs.all_attentions,)
+        return outputs
+
+    def _pad_to_mult_of_chunk_length(
+        self, input_ids, inputs_embeds, attention_mask, position_ids, input_shape, padded_seq_length, device
+    ):
+        padding_length = padded_seq_length - input_shape[-1] % padded_seq_length
+        assert (
+            self.training is False
+        ), "Sequence Length {} has to be a multiple of least common multiple chunk_length {} if training. Please consider padding the input to a length of {}.".format(
+            input_shape[-2], padded_seq_length, input_shape[-2] + padding_length
+        )
+
+        logger.info(
```

Comment:
this is gunna log every forward pass, afaict. Seems annoying.

Reply:
yeah true, I like the logging info though and it only affects when not training. Think I will leave it for now and remove later if ppl complain